### PR TITLE
[dtensor] Resolve ProcessGroup from mesh in-graph under compile_on_one_rank

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #188215

Under compile-on-one-rank (CooR) a model is traced once with make_fx into a
rank-agnostic, serializable graph. Legacy in-place c10d collectives (e.g.
dist.all_reduce, which lowers to c10d.allreduce_) take the ProcessGroup as a
torchbind ScriptObject argument, and make_fx bakes it onto the GraphModule as a
get_attr constant. torch.classes.c10d.ProcessGroup has no __getstate__, so
GraphPickler.dumps fails with "Tried to serialize object
__torch__.torch.classes.c10d.ProcessGroup which does not have a __getstate__
method defined!". The group should instead be recovered from the (graph-input)
DeviceMesh at runtime, exactly as functional collectives already do.

The fix has two parts; review device_mesh.py first, then the make_fx hook
(proxy_tensor.py) plus the shared remap in _functional_collectives.py.

1. DeviceMesh.get_group(): under CooR and an active proxy trace, return
   torch.ops._dtensor.mesh_get_process_group(self, dim) instead of a concrete
   ProcessGroup, so the group is fetched from the mesh in-graph. This mirrors
   _functional_collectives._resolve_group, which already does this for the
   functional-collective path. It is gated on get_proxy_mode() so eager
   get_group() is unchanged, and does not recurse because the op's own impl runs
   with the proxy mode disabled (get_proxy_mode() is None there).

2. get_group() alone is not sufficient for legacy in-place collectives:
   dist.all_reduce calls pg.allreduce(), and passing a pybind ProcessGroup into
   the torchbind-schema c10d.allreduce_ op mints a fresh torchbind ScriptObject
   that no longer carries make_fx's proxy slot, so the PG is re-baked and the
   mesh_get_process_group output goes dead. make_fx therefore enters a
   TorchFunctionMode (_LegacyToFunctionalCollectiveMode) under CooR that
   redirects legacy collectives to functional collectives via the existing
   traceable_collective_remaps table -- the same mapping Dynamo and non-strict
   export use. Functional collectives take the group as an op argument, which
   preserves the edge to the mesh_get_process_group node.

The redirect logic lives in a shared _remap_traceable_collective helper
(next to traceable_collective_remaps); non-strict export's
_NonStrictTorchFunctionHandler now calls it instead of carrying its own copy.

Test Plan:

New unit tests (single process, fake PG; this is the failing precompile CI
step). Confirmed they fail without the remap change.

```
python -m pytest test/distributed/tensor/test_compile_on_one_rank.py::TestCompileOnOneRankLegacyCollective -v
```

End-to-end against torchtitan's graph_trainer precompile (previously died on the
ProcessGroup serialization error; now writes the artifact with no loss-side
workaround):

```
python -m torchtitan.experiments.graph_trainer.precompile_main \
    --module graph_trainer.llama3 --config graph_trainer_llama3_debugmodel_sdpa \
    --compile.mode aot_fx_trace --compile.precompile_artifact_dir /tmp/pc \
    --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2
```

Non-strict export still redirects legacy collectives to functional ones via
the shared helper (graph has _c10d_functional.all_reduce, not c10d.allreduce_):

```
import torch, torch.distributed as dist
from torch.testing._internal.distributed.fake_pg import FakeStore
dist.init_process_group(backend="fake", store=FakeStore(), rank=0, world_size=2)
class M(torch.nn.Module):
    def forward(self, x):
        x = x.clone(); dist.all_reduce(x); return x + 1
ep = torch.export.export(M(), (torch.randn(4),), strict=False)
```

Lint:

```
lintrunner -a
```

Authored with Claude.

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -3172,6 +3172,20 @@ def _trace_inner(self, f: Callable[..., Any], *args: object) -> GraphModule:
             stack.enter_context(disable_autocast_cache())
             stack.enter_context(_set_make_fx_tracer(self))
 
+            # Under compile-on-one-rank, redirect legacy in-place c10d
```
Comment (aditvenk) [inline torch/fx/experimental/proxy_tensor.py:3192]: Can we just always do this remap? (instead of gate behind CooR)

Comment (aorenste) [reply torch/fx/experimental/proxy_tensor.py:3192]: We *could* - but the thing I'd worry about is that one of the big complaints about pytorch is when "magic" happens behind the scenes. So you ask for dist.all_reduce() and suddenly that's not what's getting called.

**Review by bobrenjc93 — APPROVED** [review]

**Comment by aorenste** [comment]
Failures are from trunk-reverted #189810 
@pytorchbot merge -i
